src/Dispatch.elm

```diff
@@ -0,0 +1,262 @@
+    , listeners'
+    , group
+    , update
+    )
+
+{-| Utility module for dispatching multiple events from a single `Html.Event`
+
+## Types
+@docs Msg
+@docs Decoder
+
+## Event handlers
+@docs listeners
+@docs listeners'
+@docs group
+
+## Dispatch
+@docs update
+@docs forward
+-}
+
+import Json.Decode as Json
+import Html.Events
+import Html.Attributes
+import Html
+import Task
+
+
+
+{-|
+  Lift any value of type `msg` to a `Cmd msg`.
+-}
+cmd : msg -> Cmd msg
+cmd msg =
+  Task.perform (always msg) (always msg) (Task.succeed msg)
+
+
+
+{-| NoOp
+ -}
+noop : Html.Attribute msg
+noop = Html.Attributes.attribute "data-dispatch-noop" "noop"
+
+{-| Message type
+-}
+type alias Msg m
+  = (List m)
+
+
+{-| A decoder with possible options
+-}
+type alias Decoder m =
+  ( Json.Decoder m, Maybe (Html.Events.Options) )
+
+
+{-| Maps messages to commands
+-}
+forward : Msg m -> Cmd m
+forward (messages) =
+  List.map cmd messages |> Cmd.batch
+
+
+{-| Map the second element of a tuple
+
+    map2nd ((+) 1) ("bar", 3) == ("bar", 4)
+-}
+map2nd : (b -> c) -> ( a, b ) -> ( a, c )
+map2nd f ( x, y ) =
+  ( x, f y )
+
+
+{-| Runs batch update
+-}
+update : (a -> b -> ( b, Cmd c )) -> Msg a -> b -> ( b, Cmd c )
+update update (msg) model =
+  let
+    inner cmd ( m, gs ) =
+      update cmd m
+        |> map2nd (flip (::) gs)
+  in
+    List.foldl
+      inner
+      ( model, [] )
+      msg
+      |> map2nd Cmd.batch
+
+
+{-| Applies given decoders to the same initial value
+   and return the applied results as a list
+-}
+applyMultipleDecoders : List (Json.Decoder m) -> Json.Decoder (List m)
+applyMultipleDecoders decoders =
+  let
+    processDecoder initial decoder =
+      Json.decodeValue decoder initial
+        |> Result.toMaybe
+  in
+    Json.customDecoder Json.value
+      (\initial ->
+        List.map (processDecoder initial) decoders
+          |> List.filterMap identity
+          |> Result.Ok
+      )
+
+
+{-|
+-}
+forwardDecoder : List (Json.Decoder a) -> Json.Decoder (Msg a)
+forwardDecoder =
+  applyMultipleDecoders
+
+
+{-| Run multiple decoders on a single Html Event
+-}
+onEvt :
+  (Msg msg -> msg)
+  -> String
+  -> List (Json.Decoder msg)
+  -> (Html.Attribute msg)
+onEvt lift event =
+  onEvtOptions lift event Html.Events.defaultOptions
+
+
+{-| Run multiple decoders on a single Html Event with
+the given options
+-}
+onEvtOptions :
+  (Msg msg -> msg)
+  -> String
+  -> Html.Events.Options
+  -> List (Json.Decoder msg)
+  -> (Html.Attribute msg)
+onEvtOptions lift event options decoders =
+  case decoders of
+    [] ->
+      noop
+
+    [ x ] ->
+      Html.Events.onWithOptions event options x
+
+    _ ->
+      forwardDecoder decoders
+        |> Json.map lift
+        |> Html.Events.onWithOptions event options
+
+
+{-| A single event
+-}
+onSingle :
+  String
+  -> Html.Events.Options
+  -> List (Json.Decoder msg)
+  -> (Html.Attribute msg)
+onSingle event options decoders =
+  case decoders of
+    [] ->
+      noop
```

Comment:
Is this better than returning `Maybe (Html.Attribtue msg)` and using `filterMap` as in the benchmark you used previously? I'd think the benchmark was better. 

Watch out, though: You sometimes do
```
   List.map f xs |> List.filterMap identity
```
which processes the list twice; better to do directly
```
   List.filterMap f xs 
```
which is equivalent.